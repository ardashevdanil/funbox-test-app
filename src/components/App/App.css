* {
  padding: 0px;
  margin: 0px;
  box-sizing: border-box;
}

.App {
  width: 100%;
  min-width: 640px;
  display: flex;
}

.points-menu {
  width: 50%;
  height: calc(100% - 20px);
  margin: 10px;
  border: solid black 2px;
  font-size: 16px;
}

.points-list {
  height: calc(100% - 43px);
  overflow-y: scroll;
}

.point {
  width: calc(100% - 20px);
  margin: 10px;
  border: solid black 2px;;
  position: relative;
  padding: 2px 3px 3px 3px;
  cursor: pointer;
  display: flex;
  justify-content: space-between;
}

.delete-button {
  height: 20px;
  width: 22px;
  left: calc(100% - 24px);
  background-color: white;
  padding: 0 5px 0 4.5px;
  font-size: 15px;
  border: inherit;
  position: absolute;
}

.delete-button:hover {
  background-color: red;
  border-color: red;
  color: white;
}

input {
  width: 100%;
  border: solid black 10px;
  font-size: 16px;
  padding: 2px;
}
