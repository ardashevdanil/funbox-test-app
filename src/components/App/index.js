import React, { Component } from 'react';

import MyMap from '../MyMap';
import PointInput from '../PointInput';
import PointsList from '../PointsList';
import './App.css';

class App extends Component {
  constructor() {
    super();
    this.state = {
      style: {},
      points: []
    };
    this.addNewPoint = this.addNewPoint.bind(this);
    this.deletePoint = this.deletePoint.bind(this);
    this.setPointCoords = this.setPointCoords.bind(this);
    this.setPointPosition = this.setPointPosition.bind(this);
  }

  addNewPoint(value) {
    const newPoint = { value: value, key: Date.now() };
    
    this.setState( (prevState) => ({
      points: prevState.points.concat(newPoint),
    }));
  }

  arrangePointsByPosition(points) {
    const sortFunc = (a, b) => {
      if (a.position < b.position) return -1;
      if (a.position > b.position) return 1;
    };
    return points.sort(sortFunc)
  }

  deletePoint(id) {
    this.setState( (prevState) => {
      const newPoints = prevState.points.filter( (item, index) => {
        if (index !== id) return true
      });
      return {points: newPoints}
    });
  }

  setPointCoords(coords, id) {
    this.setState( (prevState) => {
      const newPoints = prevState.points.map( (item, index) => {
        if (index === id) {
          item.coords = coords;
          return item
        };
        return item
      });
      return {points: newPoints}
    });
  }

  setPointPosition(position, id) {
    this.setState( (prevState) => {
      const newPoints = prevState.points.map( (item, index) => {
        if (index === id) {
          item.position = position;
          return item
        };
        return item
      });
      return {points: newPoints}
    });
  }

  componentDidMount() {
    this.setState({style: {height: document.documentElement.clientHeight}});
    window.onresize = () => {
      this.setState({style: {height: document.documentElement.clientHeight}});
    }
  }

  render() {
    const points = this.arrangePointsByPosition(this.state.points);

    return (
      <div className="App" style={this.state.style}>
        <div className="points-menu">
          <PointInput onEnterPress={this.addNewPoint}/>
          <PointsList
            onDeleteClick={this.deletePoint}
            points={points}
            setPointPosition={this.setPointPosition}
          />
        </div>
        <MyMap 
          points={this.state.points}
          setPointCoords={this.setPointCoords}
        />
      </div>
    );
  }
}

export default App;
