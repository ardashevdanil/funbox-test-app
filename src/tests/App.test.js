import React from 'react';
import Adapter from 'enzyme-adapter-react-16';
import Enzyme from 'enzyme';

import App from '../App';

Enzyme.configure({ adapter: new Adapter() });

describe('test App', () => {
	const component = Enzyme.shallow(
    <App />
  );

  test('renders correctly', () => {
    expect(component.debug()).toMatchSnapshot();
  });

  test('adds a new point', () => {
    component.instance().addNewPoint('point');
    expect(component.state('points').length).toBe(1);
    expect(component.state('points')[0].value).toEqual('point');
  });

  test('arranges points by position', () => {
    const points = [{position: 14}, {position: 0}];
    expect(component.instance().arrangePointsByPosition(points)[0].position)
      .toBe(0);
  });

  test('deletes a point', () => {
    component.setState({
      points: [{value: 'first'}, {value: 'second'}],
    });
    component.instance().deletePoint(0);
    expect(component.state('points').length).toBe(1);
    expect(component.state('points')[0].value).toEqual('second');
  });

  test('sets a point coords', () => {
    component.setState({
      points: [{ coords: [42, 24] }],
    });
    component.instance().setPointCoords([1, 25], 0)
    expect(component.state('points')[0].coords).toEqual([1, 25]);
  });
  
  test('sets a point position', () => {
    component.setState({
      points: [{ position: 42 }],
    });
    component.instance().setPointPosition(25, 0)
    expect(component.state('points')[0].position).toBe(25);
  });
})