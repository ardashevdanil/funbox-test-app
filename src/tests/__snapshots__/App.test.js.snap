// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test App renders correctly 1`] = `
"<div className=\\"App\\" style={{...}}>
  <div className=\\"points-menu\\">
    <PointInput onEnterPress={[Function: bound addNewPoint]} />
    <PointsList onDeleteClick={[Function: bound deletePoint]} points={{...}} setPointPosition={[Function: bound setPointPosition]} />
  </div>
  <MyMap points={{...}} setPointCoords={[Function: bound setPointCoords]} />
</div>"
`;
